README.md
# Mentor Interface

A REST API providing authentication and access to the Mentor Engine

## Paddle Plans, Subscriptions and Roles
Users can have different (AspNet)`Roles`, e.g. "Admin" or "Premium", which may be required for accessing certain resources.
Some Roles are available to users through `Subscriptions`.
For each `Subscription` there may be one or multiple PaddlePlans the user can choose from, differing e.g. by contract duration and pricing. 

## Persisting Cookies
To persist cookies, mount a `PVC` to `/keys` - This way MentorInterface can remount the keys to validate existing cookies upon relaunch.

## Environment Variables

- `STEAM_API_KEY` : Steam application key for OpenID authentication. [\*]
- `MYSQL_CONNECTION_STRING` : Connection string for the User Database. [\*]
- `IS_MIGRATING` : Warning: Do not use in production. If this is set to true, MentorInterface will be able to run migrations but nothing else. Only works if `MYSQL_CONNECTION_STRING` is also provided.
- `IDENTITY_WORKAROUND_BEARER_TOKEN` : The token which devs need to add in Headers["Authorization"]="Bearer [token]" to authenticate
- `IDENTITY_WORKAROUND_USER_ID` : integer, the default ApplicationUserId of the User assigned to devs who login with above Bearer Token. Can be overriden by specifying Headers["Impersonate-ApplicationUserId"]
- `PADDLE_VENDOR_ID` : int, provided by Paddle for identifying us. [\*]
- `PADDLE_VENDOR_AUTH_CODE` : string, our secret key for communication with Paddle.[\*]

[\*] *Required*


## Monitoring / Prometheus

Collect metrics on port `9913` (Defined in `Startup.cs:METRICS_PORT`) at `/metrics`


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

# [1.12.0] - 2020-08-13
### Changed
- Identity and Correlation Cookies to `Secure` 

# [1.11.2] - 2020-07-24
### Added
- `AcceptedTermsOfService` Field
    - Accept TOS On new Paddle Subscription

# [1.11.1] - 2020-07-23
### Added
- Ability to persist cookies over restarts

## [1.11.0] - 2020-07-22
### Changed
- Remove authentication requirement for getting Subscription information
### Added
- IdentityController for multiple SteamIds
- Situation Feedback endpoint

## [1.10.0] - 2020-06-23
### Changed
- Controllers for dashboard endpoints to return full data when demo profile is visited

## [1.9.0] - 2020-06-18
### Added
- Env var `STEAMUSEROPERATOR_URL_OVERRIDE`
- Influencer SubscriptionType, checked on-the-fly based on whether non-paying users have "mentor.gg" in their name

## [1.8.0] - 2020-06-17
### Added
- More Situation Operator endpoints for debugging
- `forceRefresh` param to PlayerInfo endpoint

## [1.7.0] - 2020-06-02
### Added
- More Situation Operator Routes

## [1.6.0] - 2020-05-29
### Added
- Situation Operator Routes

## [1.5.1] - 2020-05-15
### Changed
- Internal DNS lookup for connected services no longer limited to `default` namespace.
- Increase ApplicationCookie `MentorInterface.Identity` expiry time to `365` Days.

## [1.5.0] - 2020-05-06
### Changed
- Add DownloadController with endpoint for providing  blobUrls

## [1.4.1] - 2020-05-01
### Changed
- /subscriptions now returns all available subscriptions, not just those available to up/downgrade for the specific user

## [1.4.0] - 2020-04-30
### Changed
- Send `SubscriptionType` to `MatchRetriever` instead of `DailyMatchesLimit`

## [1.3.0] - 2020-04-03
### Added
- Option for devs to specify Header "Impersonate-ApplicationUserId" to login as specific user

## [1.2.0] - 2020-04-03
### Changed
- Raised daily matches limit for Premium users from 5 per day to unlimited

## [1.1.0] - 2020-04-03
### Changed
- `/valve/look/{steamId}` Endpoint now passes Low Quality by default.
- `/faceit/look/{steamId}` Endpoint now passes Low Quality by default.

## [1.0.1] - 2020-04-03
### Changed
- PaddlePublicKey.pem

## [1.0.0] - 2020-04-03
### Added
- Anonymous route for user Identity only accessiable internal at `/identity/{steamId}`.

### Changed
- `/single/{steamId}/demostatus/failed-demos` now considers the SteamId.

## [0.2.4] - 2020-03-19
### Added
- PaddleApiCommunicator
- SubscriptionRemover
- Environment Variables: PADDLE_VENDOR_ID, PADDLE_VENDOR_AUTH_CODE

### Changed
- Implementation of Subscriptions (regarding Paddle & the webapp)

## [0.2.3] - 2020-03-13
### Added
- Logic for computing DailyMatchesLimit
- Controller for handling subscriptions
- Workaround for making devs seem authenticated as a User when accessing from localhost where the IdentityCookie can not be used
- Environment vars IDENTITY_WORKAROUND_USER_ID and IDENTITY_WORKAROUND_BEARER_TOKEN for above workaround

### Changed
- PaddlePlans and Roles are now created differently and coupled tighter than before
- Cleared all migrations and start with fresh ones


## [0.2.2] - 2020-03-11
### Changed
- Bugfix: Seed Database after migrations

## [0.2.1] - 2020-03-11
### Added
- Database migration on startup.

## [0.2.0] - 2020-03-11
### Added
- Console Timestamps.
- Paddle Webhook Intergration.
- Routes for Match Retreiver and Demo Central.
- Development CORS.
### Changed
- Rename env var MYSQL_CONNECTIONSTRING to MYSQL_CONNECTION_STRING.
- Fix Signout endpoint.

## [0.1.1] - 2020-01-24
### Added
- Routes for FaceItMatchGatherer and SharingCodeGatherer.
- MySQL Connections retries.
### Changed
- Net Core Framework to 3.1.
### Removed
- Let's Encrpyt SSL (SSL is now configured with the Ingress on K8s).


## [0.1.0] - 2020-01-02 
### Added 
- Swagger documentation.
- Identity authentication (Steam).
- User database.
- Let's Encrypt SSL.


MentorInterface/Startup.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;
using Microsoft.AspNetCore.Authentication.Cookies;
using MentorInterface.Authentication;
using MentorInterface.Middleware;
using Microsoft.OpenApi.Models;
using System.Reflection;
using System.IO;
using Microsoft.EntityFrameworkCore;
using Entities.Models;
using Database;
using Prometheus;
using MentorInterface.Helpers;
using Microsoft.AspNetCore.Identity;
using Entities.Models.Paddle;
using MentorInterface.Paddle;
using System.Security.Claims;
using Microsoft.AspNetCore.Authentication;
using System.Net.Http;
using Microsoft.AspNetCore.DataProtection;
using Microsoft.AspNetCore.Http;

namespace MentorInterface
{
    /// <summary>
    /// Congifure the WebApi's enviroment.
    /// </summary>
    public class Startup
    {

        private bool IsDevelopment => Configuration.GetValue<string>("ASPNETCORE_ENVIRONMENT") == Environments.Development;

        /// <summary>
        /// The token which devs need to add in Headers["Authorization"]="Bearer <token>" to authenticate
        /// </summary>
        private string IDENTITY_WORKAROUND_BEARER_TOKEN => Configuration.GetValue<string>("IDENTITY_WORKAROUND_BEARER_TOKEN");

        /// <summary>
        /// The ApplicationUserId of the User assigned to devs who login with above Bearer Token.
        /// </summary>
        private string IDENTITY_WORKAROUND_USER_ID => Configuration.GetValue<string>("IDENTITY_WORKAROUND_USER_ID");

        /// <summary>
        /// Amount of times to attempt a successful MySQL connection on startup.
        /// </summary>
        const int MYSQL_RETRY_LIMIT = 3;

        /// <summary>
        /// Port to scrape metrics from at `/metrics`
        /// </summary>
        public const int METRICS_PORT = 9913;

        /// <summary>
        /// Extend and apply a supplied configuration.
        /// </summary>
        /// <param name="configuration">Configuration</param>
        public Startup(IConfiguration configuration)
        {
            Configuration = new ConfigurationBuilder()
                .AddConfiguration(configuration)
                .AddEnvironmentVariables()
                .Build();
        }

        /// <summary>
        /// Current configuration.
        /// </summary>
        public IConfiguration Configuration { get; }


        /// <summary>
        /// Add services to the container.
        /// </summary>
        /// <param name="services"></param>
        public virtual void ConfigureServices(IServiceCollection services)
        {
            services.AddControllers()
                .AddNewtonsoftJson(x => 
                {
                    // Serialize JSON using the Member CASE!
                    x.UseMemberCasing();
                    // Serialize longs (steamIds) as strings
                    x.SerializerSettings.Converters.Add(new LongToStringConverter());
                });
            services.AddApiVersioning();

            #region Identity

            // Connect to the user database.
            var connString = Configuration.GetValue<string>("MYSQL_CONNECTION_STRING");
            if (connString != null)
            {
                services.AddDbContext<ApplicationContext>(o =>
                {
                    o.UseMySql(connString, sqlOptions =>
                    {
                        sqlOptions.EnableRetryOnFailure(MYSQL_RETRY_LIMIT);
                    });

                }, ServiceLifetime.Scoped);
            }
            // Use InMemoryDatabase if the connectionstring is not set in a DEV enviroment
            else if (IsDevelopment)
            {
                services.AddDbContext<ApplicationContext>(o =>
                {
                    o.UseInMemoryDatabase("MyTemporaryDatabase");

                }, ServiceLifetime.Scoped);

                Console.WriteLine("WARNING: Using InMemoryDatabase!");
            }
            else
            {
                throw new ArgumentException(
                    "MySqlConnectionString is missing, configure the `MYSQL_CONNECTION_STRING` enviroment variable.");
            }
            
            if (Configuration.GetValue<bool>("IS_MIGRATING"))
            {
                Console.WriteLine("IS_MIGRATING is true! ARE YOU STILL MIGRATING?");
                return;
            }

            services.AddIdentity<ApplicationUser, ApplicationRole>()
                .AddEntityFrameworkStores<ApplicationContext>();

            #endregion

            #region Logging
            services.AddLogging(o =>
            {
                o.AddConsole(o =>
                {
                    o.TimestampFormat = "[yyyy-MM-dd HH:mm:ss zzz] ";
                });

                // Filter out ASP.Net and EFCore logs of LogLevel lower than LogLevel.Warning
                o.AddFilter("Microsoft.EntityFrameworkCore.Database.Command", LogLevel.Warning);
                o.AddFilter("Microsoft.EntityFrameworkCore.Infrastructure", LogLevel.Warning);
                o.AddFilter("Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker", LogLevel.Warning);
                o.AddFilter("Microsoft.AspNetCore.Mvc.Infrastructure.ObjectResultExecutor", LogLevel.Warning);
                o.AddFilter("Microsoft.AspNetCore.Hosting.Diagnostics", LogLevel.Warning);
                o.AddFilter("Microsoft.AspNetCore.Routing.EndpointMiddleware", LogLevel.Warning);

                // Filter logs for each forwarded request
                o.AddFilter("System.Net.Http.HttpClient", LogLevel.Warning);
            });
            #endregion

            #region Helpers
            services.AddTransient<IRoleHelper, RoleHelper>();
            services.AddTransient<ISteamUserOperator, SteamUserOperator>();
            #endregion

            #region Paddle & Subscriptions

            services.AddTransient<IWebhookVerifier, WebhookVerifier>(x =>
                {
                    return new WebhookVerifier(
                        File.ReadAllText("/app/Paddle/PaddlePublicKey.pem"));
                }
            );

            var PADDLE_VENDOR_ID = Configuration.GetValue<int>("PADDLE_VENDOR_ID");
            var PADDLE_VENDOR_AUTH_CODE = Configuration.GetValue<string>("PADDLE_VENDOR_AUTH_CODE")
                ?? throw new ArgumentNullException("The environment variable PADDLE_VENDOR_AUTH_CODE has not been set.");

            services.AddTransient<IPaddleApiCommunicator, PaddleApiCommunicator>(sp =>
            {
                return new PaddleApiCommunicator(sp.GetRequiredService<ILogger<PaddleApiCommunicator>>(), sp.GetRequiredService<IHttpClientFactory>(), PADDLE_VENDOR_ID, PADDLE_VENDOR_AUTH_CODE);
            });

            services.AddTransient<SubscriptionRemover>();
            services.AddHostedService<SubscriptionRemoverBackgroundService>();
            #endregion

            #region HTTP Clients

            // Add HTTP clients for communication with other services in the cluster
            services.AddConnectedHttpService(ConnectedServices.DemoCentral, Configuration, "DEMOCENTRAL_URL_OVERRIDE");
            services.AddConnectedHttpService(ConnectedServices.FaceitMatchGatherer, Configuration, "FACEITMATCHGATHERER_URL_OVERRIDE");
            services.AddConnectedHttpService(ConnectedServices.MatchRetriever, Configuration, "MATCHRETRIEVER_URL_OVERRIDE");
            services.AddConnectedHttpService(ConnectedServices.SharingCodeGatherer, Configuration, "SHARINGCODEGATHERER_URL_OVERRIDE");
            services.AddConnectedHttpService(ConnectedServices.SituationOperator, Configuration, "SITUATIONOPERATOR_URL_OVERRIDE");
            services.AddConnectedHttpService(ConnectedServices.SteamUserOperator, Configuration, "STEAMUSEROPERATOR_URL_OVERRIDE");

            // Add HTTP clients for external communication
            services.AddConnectedHttpService(ConnectedServices.PaddleApi, Configuration, "PADDLEAPI_URL_OVERRIDE");
            #endregion

            #region Application Cookie
            services
                .ConfigureApplicationCookie(options =>
                {
                    options.Cookie.Name = "MentorInterface.Identity";
                    options.Cookie.IsEssential = true;
                    options.Cookie.SameSite = SameSiteMode.None;
                    options.Cookie.SecurePolicy = CookieSecurePolicy.Always;
                    options.LoginPath = "/authentication/signin/steam";
                    options.LogoutPath = "/authentication/signout/steam";

                    // Set the Cookie lifetime to ONE year.
                    options.ExpireTimeSpan = TimeSpan.FromDays(365);

                    // Return 401 OR 403 instead of redirecting to LoginPath
                    options.Events = new CookieAuthenticationEvents
                    {
                        OnRedirectToLogin = async (context) => {
                            context.Response.StatusCode = 401;
                            await Task.CompletedTask;
                        },

                        OnRedirectToAccessDenied = async (context) => {
                            context.Response.StatusCode = 403;
                            await Task.CompletedTask;
                        },
                    };
                });
            #endregion

            #region Steam Authentication
            // Load the Authentication Section and confirm the entry exists.
            var steamApplicationKey = Configuration.GetValue<string>("STEAM_API_KEY");
            if (steamApplicationKey != null)
            {
                services
                    .AddAuthentication()
                    .AddSteam(scheme: MentorAuthenticationSchemes.STEAM, options =>
                    {
                        options.CorrelationCookie.SameSite = SameSiteMode.None;
                        options.CorrelationCookie.SecurePolicy = CookieSecurePolicy.Always;
                        options.ApplicationKey = steamApplicationKey;
                        options.CallbackPath = "/openid/callback/steam";
                    });
            }
            // If in a DEV context, warn, dont throw.
            else if (IsDevelopment)
            {
                Console.WriteLine("WARNING: `STEAM_API_KEY` is missing!");
            }
            else
            {
                throw new ArgumentException("SteamApplicationKey is missing, configure the `STEAM_API_KEY` enviroment variable.");
            }
            #endregion


            services.AddDataProtection()
                    .SetApplicationName($"mentor-interface")
                    .PersistKeysToFileSystem(new DirectoryInfo($@"/keys"));

            #region Cors
            services.AddCors(o => o.AddPolicy("Debug", builder =>
            {
                var allowedOrigins = new string[]
                {
                    "http://localhost:8080",
                    "https://localhost:8080",
                };
                builder.WithOrigins(allowedOrigins)
                    .AllowAnyMethod()
                    .AllowAnyHeader()
                    .AllowCredentials();
            }));
            #endregion

            #region Swagger
            services.AddSwaggerGen(options =>
            {
                OpenApiInfo interface_info = new OpenApiInfo { Title = "Mentor Interface", Version = "v1", };
                options.SwaggerDoc("v1", interface_info);

                // Generate documentation based on the Docstrings provided.
                var xmlFile = $"{Assembly.GetExecutingAssembly().GetName().Name}.xml";
                var xmlPath = Path.Combine(AppContext.BaseDirectory, xmlFile);
                options.IncludeXmlComments(xmlPath);

                options.EnableAnnotations();
            });
            #endregion

        }

        /// <summary>
        /// Configure the HTTP request pipeline.
        /// </summary>
        /// <param name="app"></param>
        /// <param name="env"></param>
        public virtual void Configure(IApplicationBuilder app, IWebHostEnvironment env, IServiceProvider serviceProvider)
        {
            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
            }

            if (IsDevelopment)
            {
                app.UseCors("Debug");
            }

	    app.UseMiddleware(typeof(ErrorHandlingMiddleware));

            #region Swagger
            app.UseSwagger();

            app.UseSwaggerUI(options =>
            {
                options.RoutePrefix = "swagger";
                options.SwaggerEndpoint("/swagger/v1/swagger.json", "Mentor Interface");
            });
            #endregion

            app.UseRouting();

            // Who is the user?
            app.UseAuthentication();

            // Configure the app to check for an authorization header with the configured bearer token
            // If it is present, authenticate as the user with the configured User Id
            // The entire app.Use() statement could be removed while keeping the Cookie-based login intact
            app.Use(async (context, next) =>
            {
                // Check if authorization header for workaround login is available
                if(!context.User.Identity.IsAuthenticated && IDENTITY_WORKAROUND_BEARER_TOKEN != null && context.Request.Headers["Authorization"] == "Bearer " + IDENTITY_WORKAROUND_BEARER_TOKEN)
                {
                    var principal = new ClaimsPrincipal();

                    // Determine Application User Id
                    // Authenticate as user specified by header or use default value instead
                    string appUserId = long.TryParse(context.Request.Headers["Impersonate-ApplicationUserId"], out var applicationUserId) 
                        ? applicationUserId.ToString()
                        : IDENTITY_WORKAROUND_USER_ID;

                    context.User = principal;

                    // Assign the current user the identity of the user with the User Id IDENTITY_WORKAROUND_USER_ID
                    var claims = new List<Claim>();
                    claims.Add(new Claim(ClaimTypes.NameIdentifier, appUserId));
                    var claimsIdentity = new ClaimsIdentity(claims, "Identity.Application");
                    principal.AddIdentity(claimsIdentity);
                }

                await next();
            });

            // Is the user allowed to perform this action?
            app.UseAuthorization();

            app.UseEndpoints(endpoints =>
            {
                endpoints.MapControllers();
            });

            app.UseMetricServer(METRICS_PORT);

            // Migrate if this is not an inmemory database
            if (serviceProvider.GetRequiredService<ApplicationContext>().Database.ProviderName != "Microsoft.EntityFrameworkCore.InMemory")
            {
                serviceProvider.GetRequiredService<ApplicationContext>().Database.Migrate();
            }

            SeedDatabase(serviceProvider);
        }

        /// <summary>
        /// Seed the Database with Role information.
        /// </summary>
        /// <param name="serviceProvider"></param>
        private void SeedDatabase(IServiceProvider serviceProvider)
        {
            RoleCreator.CreateRoles(serviceProvider, RoleCreator.ApplicationRoles);

            // Write PaddlePlans to db and connect them with Roles
            var plans = PaddlePlanManager.ProductionPlans;
            PaddlePlanManager.SetPaddlePlans(serviceProvider, plans);
        }
    }
}
